### PR TITLE
feat(Options) - Annotate the record in the database with the source o…

### PR DESCRIPTION
**Pull request by owensk**
In the options automator project (which is making options update self service and also supported in multiple regions) we are adding a new way to set options based on a git repo that is synchronized with the options backing storage. 
This allows for: accountability, change review for the options that need it, support for multiple environments (region override).

Right now there is a number of channels an option can be changed through: CLI, killswitch CLI, the admin UI, the application itself. 
Since we are moving the source of truth outside of the DB and the existing channels will still exists (some of them only temporarily, others will be emergency channels and other will be restricted to specific types of options) drift can happen.

This PR adds an `last_updated_by` column in the options table so we can detect drift when it happens in the process of introducing the new system.
It will be used by preventing the git defined options from overriding the newer updates made by someone else.

### CODE REVIEW

**Review by kneeyo1 — DISMISSED** [review]

File: src/sentry/options/manager.py
Comment (markstory) [inline src/sentry/options/manager.py:80]: My guess is cyclic imports, or Django getting mad that you're using models before the application is 'ready'. For the cyclic imports you can hack around that using:

```python
if typing.TYPE_CHECKING:
   from sentry.models.options.option import OptionsType
```

And then defining the type hints as strings.

**Review by markstory — COMMENTED** [review]
Looks ok. The test failures seem relevant though.

File: src/sentry/options/manager.py
Comment (wedamija) [reply src/sentry/options/manager.py:80]: Or `from __future__ import annotations` and you can skip the strings

File: src/sentry/models/options/option.py
Comment (wedamija) [inline src/sentry/models/options/option.py:39]: Will we need to query by this at all?

File: src/sentry/options/manager.py
Comment (wedamija) [inline src/sentry/options/manager.py:288]: Tbh we could just have the manager and model in the same file. Otherwise, you could also just import it in this function instead of the top level.

File: src/sentry/models/options/option.py
Comment (owensk) [reply src/sentry/models/options/option.py:39]: I don't anticipate writing queries against this column directly. This is specifically for my config options tooling that wants to know who set the most recent value. I can't think of a useful instance where we need to query by the last_update_by column. If we did, there are only ~300 options entries so it's a pretty small table all things considered.

File: src/sentry/options/manager.py
Comment (owensk) [reply src/sentry/options/manager.py:80]: Cool. I'll give those a shot. Thanks!

File: src/sentry/models/options/option.py
Comment (fpacifici) [inline src/sentry/models/options/option.py:15]: Should we have an entry for the killswitch system as well ?
Not a blocking concern

File: src/sentry/options/manager.py
Comment (fpacifici) [inline src/sentry/options/manager.py:81]:
```
# TODO: Find out why `from sentry.models.options import OptionsTypes` doesn't work at top level
```

What problem are you running into ?

**Review by fpacifici — COMMENTED** [review]
Please see the test failures. Several are relevant 
```
TypeError: make_key() missing 1 required positional argument: 'source'
```

File: src/sentry/models/options/option.py
Comment (owensk) [reply src/sentry/models/options/option.py:15]: We definitely should. I'll add one in there.

File: src/sentry/options/manager.py
Comment (owensk) [reply src/sentry/options/manager.py:81]: I believe it's an cyclical import or other import dependency that's causing this to not work. Some discussion [here](https://github.com/getsentry/sentry/pull/47510/files#r1171685519) that's related. I think I may pull the enum into the options module and then import from there into the models module. That's just an idea so I'd be interested in hearing others.

**Comment by corps** [comment]
Hi @owensk , I'm working on splitting the `Option` model into two identical models that subclass an abstract superclass here https://github.com/getsentry/sentry/pull/48833.

Are you planning on wrapping up this PR in the near future?  If not, I can merge my changes and help merge in those changes into this branch.  If you are nearing wrapping up this PR, I can hold off until you've landed to keep things easier for you.

**Comment by corps** [comment]
I'll update and fix conflicts on this branch soon.

File: src/sentry/models/options/option.py
Comment (wedamija) [inline src/sentry/models/options/option.py:37]: This `max_length` probably needs to be reflected in the migration state and the ddl

File: src/sentry/options/manager.py
Comment (evanpurkhiser) [inline src/sentry/options/manager.py:27]: might be nice to have docstring comments on some of these

Comment (fpacifici) [reply src/sentry/options/manager.py:27]: right, makes sense

File: src/sentry/options/manager.py
Comment (cmanallen) [inline src/sentry/options/manager.py:147]: Since you're touching this method there's a good opportunity to add type hinting.  E.g. we know it returns key so `-> Key:` could be applied here.  If the other parameters have well known values you could add type hinting there as well.

**Review by kneeyo1 — APPROVED** [review]

File: src/sentry/options/store.py
Comment (wedamija) [reply src/sentry/options/store.py]: That looks correct to me. Where are you seeing this error? Tests/locally/etc?

Comment (fpacifici) [reply src/sentry/options/store.py]: Works for me as well. 
@kneeyo1  mind checking that the migration happened on your machine ?
```
sentry django showmigrations
```

File: src/sentry/options/manager.py
Comment (fpacifici) [reply src/sentry/options/manager.py:161]: I added types around the manager module. Could not simply add the return type here because this module cannot depend on the module that defines the Key. I'd have to add the typehint as a string.

File: src/sentry/options/manager.py
Comment (fpacifici) [reply src/sentry/options/manager.py]: done

**Review by markstory — APPROVED** [review]
Schema changes look good to me.

**Review by wedamija — APPROVED** [review]

File: src/sentry/options/store.py
Comment (kneeyo1) [reply src/sentry/options/store.py]: Migration looks good to me

**Pull request merged**
